validator/src/consensus/signing/client.ts

```diff
@@ -0,0 +1,197 @@
+import { encodePacked, type Hex, keccak256 } from "viem";
+import type { GroupId, ParticipantId, SignatureId } from "../../frost/types.js";
+import { generateMerkleProofWithRoot } from "../merkle.js";
+import type {
+	GroupInfoStorage,
+	SignatureRequestStorage,
+	SigningCoordinator,
+} from "../types.js";
+import { groupChallenge, lagrangeCoefficient } from "./group.js";
+import {
+	bindingFactors,
+	calculateGroupCommitment,
+	createNonceTree,
+	decodeSequence,
+	groupCommitementShares,
+	nonceCommitmentsWithProof,
+	type PublicNonceCommitments,
+} from "./nonces.js";
+import { createSignatureShare, lagrangeChallenge } from "./shares.js";
+import { verifySignatureShare } from "./verify.js";
+
+export class SigningClient {
+	#storage: GroupInfoStorage & SignatureRequestStorage;
+	#coordinator: SigningCoordinator;
+
+	constructor(
+		storage: GroupInfoStorage & SignatureRequestStorage,
+		coordinator: SigningCoordinator,
+	) {
+		this.#storage = storage;
+		this.#coordinator = coordinator;
+	}
+
+	async commitNonces(groupId: GroupId) {
+		const signingShare = this.#storage.signingShare(groupId);
+		if (signingShare === undefined) throw Error(`No info for ${groupId}`);
+		const nonceTree = createNonceTree(signingShare);
+		const nonceTreeHash = this.#storage.registerNonceTree(nonceTree);
+		await this.#coordinator.publishNonceCommitmentsHash(groupId, nonceTreeHash);
+	}
+
+	async handleNonceCommitmentsHash(
+		groupId: GroupId,
+		senderId: ParticipantId,
+		nonceCommitmentsHash: Hex,
+		chunk: bigint,
+	) {
+		const participantId = this.#storage.participantId(groupId);
+		// Only link own nonce commitments
+		if (participantId !== senderId) return;
+		this.#storage.linkNonceTree(groupId, chunk, nonceCommitmentsHash);
+	}
+
+	async handleSignatureRequest(
+		groupId: GroupId,
+		signatureId: SignatureId,
+		message: Hex,
+		sequence: bigint,
+	) {
+		// TODO: check if we really want to sign the message
```

Comment:
What, no blind signing? I think we should invent a protocol so you can blind sign this `message` with confidence. Maybe called "Schildnetz"?